rpc/src/rpc_service.rs

```diff
@@ -278,15 +283,23 @@ impl RequestMiddleware for RpcRequestMiddleware {
                     .body(hyper::Body::from(self.health_check()))
                     .unwrap()
                     .into(),
-                "/metrics" => hyper::Response::builder()
-                    .status(hyper::StatusCode::OK)
-                    .header("Content-Type", "text/plain; version=0.0.4; charset=UTF-8")
-                    .body(hyper::Body::from(render_prometheus(
-                        &self.bank_forks,
-                        &self.health.cluster_info,
-                    )))
-                    .unwrap()
-                    .into(),
+                "/metrics" => {
+                    let rpc_processor = self.rpc_processor.as_ref().unwrap();
```

Comment:
So I looked into the code for `BankForks` a bit, it looks like `BankForks` contains almost everything we need! We can do `.get(slot)` to get the bank for a particular slot. It somehow feels wrong to pass in `JsonRpcRequestProcessor` in the middleware that goes in front of the `JsonRpcRequestProcessor` ...

There is `BlockCommitmentCache` in `JsonRpcRequestService`. It looks like that stores the number of confirmations per block, and it can give us the slot for a certain commitment level. It is available when the `RpcRequestMiddleware` is constructed, so we can pass in that commitment cache, instead of a full `JsonRpcRequestService`. From there we can get the slots:

https://github.com/solana-labs/solana/blob/d9c259a2313f620b4b64b7cec0d55dd4234fbcd9/runtime/src/commitment.rs#L115-L125

For the block height metrics, we wouldn’t even need to get that slot’s bank and read its clock, we would get the slot directly.

<sub>Interesting aside, so it looks like Solana’s `BankForks` can only store a single bank per slot. So if there is a fork, there is no way for it to track multiple forks simultaneously. It would have to throw out all progress on one fork if it decides to switch to another.</sub>